Refactor `bootstrap`, `bootstrap-conda`, `m4/sage_spkg_collect.m4`, `sage-spkg-info` through `sage --package properties`, `sage-get-system-packages`

mkoeppe opened the pull request:
<!-- ^^^^^
Please provide a concise, informative and self-explanatory title.
Don't put issue numbers in there, do this in the PR body below.
For example, instead of "Fixes #1234" use "Introduce new method to calculate 1+1"
-->
<!-- Describe your changes here in detail -->

Using the command `sage --package properties` added in https://github.com/sagemath/sage/pull/37018, we eliminate some direct references to `build/pkgs/` from various scripts and reduce code duplication in reading SPKG metadata.

Using other `sage --package` commands, we also simplify the code for `bootstrap -s` and `bootstrap -D`.

Review instructions: Note that there are changes to the `bootstrap`-generated file `m4/sage_spkg_configures.m4`: 
- The macro `SAGE_SPKG_COLLECT_INIT` is now called explicitly, with an argument (the full list of packages).
- Lines such as `m4_define([SPKG_INSTALL_REQUIRES_exceptiongroup], [['exceptiongroup; python_version<\"3.11\"',]])dnl` are added for use by the macro `SAGE_PYTHON_PACKAGE_CHECK` (replacing direct access to `version_requirements.txt` files) 

<!-- Why is this change required? What problem does it solve? -->
<!-- If this PR resolves an open issue, please link to it here. For example "Fixes #12345". -->
This is:
- Split out from #36740
<!-- If your change requires a documentation PR, please link it appropriately. -->

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!-- Feel free to remove irrelevant items. -->

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on
- #12345: short description why this is a dependency
- #34567: ...
-->

<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

Commit:
m4/sage_spkg_collect.m4: Handle wheel packages like normal packages

Commit:
m4/sage_spkg_collect.m4: Use 'sage-package properties' to get versions

Commit:
m4/sage_spkg_collect.m4: Fix handling of missing package dirs (test: './bootstrap && rm -rf build/pkgs/cachetools; ./configure')

Commit:
Repair 'bootstrap -D'

Commit:
bootstrap: Use 'sage -package update' to update the configure tarball version/checksums

Commit:
bootstrap -D: Simplify using 'sage-package download'

tobiasdiez commented on bootstrap-conda:
```diff
@@ -18,10 +18,11 @@ SAGELIB_PACKAGES=
 SAGELIB_OPTIONAL_PACKAGES=
 DEVELOP_PACKAGES=
 
+eval $(sage-package properties --format=shell :all:)
```
I don't think it's a good idea to tie bootstrap conda even closer to the sage-the-distribution scripts. Please revert the changes in the bootstrap-conda file.

mkoeppe replied:
Nonsense

saraedum replied:
@mkoeppe could you elaborate why this is "nonsense" from your point of view?

mkoeppe replied:
This PR replaces direct access to the files in `build/pkgs` (an implementation detail of the sage-bootstrap package) by access through a script of the sage-bootstrap package. Such refactoring does not "tie bootstrap closer to sage-the-distribution scripts".

Commit:
bootstrap: Remove use of CONFVERSION

Commit:
bootstrap: Simplify by using 'sage-package properties'

bootstrap: Remove last direct use of 'build/pkgs'

bootstrap-conda: Use 'sage-package properties' and 'sage-package dependencies'

Update shell uses of 'sage-package properties'

Commit:
m4/sage_python_package_check.m4: Replace direct access to version_requirements.txt

Commit:
build/bin/sage-get-system-packages: Restore correct search order for 'SYSTEM=pip'

Commit:
build/bin/sage-spkg-info: Show versions using new command 'sage-get-system-packages versions'

Commit:
m4/sage_python_package_check.m4: Update doc to mention src/pyproject.toml

Commit:
m4/sage_spkg_collect.m4: Repair dependencies

Commit:
build/bin/sage-spkg-info: Fix rst format

Commit:
build/bin/sage-get-system-packages versions: Fix for jupyter_core, broken in #36982 + #37894

kwankyu commented on m4/sage_python_package_check.m4:
```diff
@@ -19,8 +19,9 @@
+#     5. If the contents of version_requirements.txt (or entry in
+#        "src/pyproject.toml")are not met (wrong version, no version,
+#        etc.) by the system python.
 #
```
```
) are
```

kwankyu reviewed (approved):
Thanks. Working well.

Commit:
m4/sage_python_package_check.m4: Fix typo in comment

Commit:
bootstrap: Fix -u
